adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -463,17 +473,103 @@ private void handleConnectionRequestResult(final MqttEndpoint endpoint,
 
+        handleCommandSubscription(endpoint, authenticatedDevice);
+
         endpoint.closeHandler(v -> {
             close(endpoint, authenticatedDevice);
-            LOG.debug("connection to device [tenant-id: {}, device-id: {}] closed",
-                    authenticatedDevice.getTenantId(), authenticatedDevice.getDeviceId());
-            metrics.decrementMqttConnections(authenticatedDevice.getTenantId());
         });
 
         endpoint.publishHandler(message -> handlePublishedMessage(new MqttContext(message, endpoint, authenticatedDevice)));
         metrics.incrementMqttConnections(authenticatedDevice.getTenantId());
     }
 
+    private void handleCommandSubscription(final MqttEndpoint endpoint, final Device authenticatedDevice) {
+        endpoint.subscribeHandler(subscribe -> {
+            final List<MqttQoS> grantedQosLevels = new ArrayList<>();
+            if (subscribe.topicSubscriptions().size() == 1) {
+                final MqttTopicSubscription subscription = subscribe.topicSubscriptions().get(0);
+                final CommandSubscription cmdSub = CommandSubscription.fromTopic(subscription.topicName(),
+                        authenticatedDevice);
+                if (cmdSub != null) {
+                    if (subscription.qualityOfService() != MqttQoS.EXACTLY_ONCE) {
+                        LOG.debug("subscription for [topic:{}, QoS:{}]", subscription.topicName(),
+                                subscription.qualityOfService());
+                        grantedQosLevels.add(subscription.qualityOfService());
+                        endpoint.subscribeAcknowledge(subscribe.messageId(), grantedQosLevels);
+                        createCommandConsumer(endpoint, cmdSub).setHandler(h -> {
+                            if (h.succeeded()) {
+                                sendTtdEvent(cmdSub.getTenant(), cmdSub.getDeviceId(), authenticatedDevice, -1);
+                            } else {
+                                LOG.error("command consumer could not be created for [tenant:{}, device:{}]: {}",
+                                        cmdSub.getTenant(), cmdSub.getDeviceId(), h.cause());
+                            }
+                        });
+                        // if called from unauthenticated device subscription overwrite the close
+                        // handler with now known tenant and device
+                        if (authenticatedDevice == null) {
+                            endpoint.closeHandler(v -> {
+                                close(endpoint, new Device(cmdSub.getTenant(), cmdSub.getDeviceId()));
+                            });
+                        }
+                    } else {
+                        LOG.error("denied subscription for exactly once");
+                    }
+                }
+            } else {
+                LOG.error("subscription for multiple topics is not allowed");
+            }
+            if (grantedQosLevels.isEmpty()) {
```

Comment:
Feels akward to me to use the `grantedQosLevels` as the indicator if all went ok with the subscription.
Wouldn't an explicit boolean be better for this?
As far as I can see the grantedQosLevels is only used at one line and could therefore be eliminated (if using a boolean for the other purpose it currently has, namely indicating the outcome of a subscription).
WDYT?

Reply:
The indication of the unsuccessful subscription is also done with the grantedQoSLevel "MqttQos.FAILURE" in proton then. I think this is a good match.

Reply:
Ok, then this makes more sense to me. But it might be worth a comment, since you need this specific knowledge...